### PR TITLE
To Do List Add & Remove

### PR DESCRIPTION
**Pull request by GraceAbike**
# To Do list: Add, Remove & Edit.
In this project, I built a simple HTML list of To Do tasks. The list was styled according to the specifications listed in this lesson. This simple web page was built using webpack and served by a webpack dev server.
and also I implemented the CRUD (create, read, update, delete) methods. All the elements of the user interface is fully functional and my application will is complete.

> Learning objectives

complete-website-with-nav
- To Use webpack to bundle JavaScript.
- To Learn how to use proper ES6 syntax.
- To Use ES6 modules to write modular JavaScript.

## Built With

- HTML
- CSS
- VANILLA JAVASCRIPT
- WEB PACK

## Additional Tools

- WebHint
- ESlint
- Stylelint
- Node Packet Manager
- Git and GitHub

## Getting Started

To get a local copy of this project up and running on your local machine follow the steps below.

### Prerequisites

- Basic html, css and medium Javascript know how.
- Git installation
- A text editor
- A terminal
- A web browser to view output e.g Google Chrome
- An IDE e.g Visual studio code

### CODE REVIEW

File: src/modules/addTask.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      renderToDoItems();
+      todoinputEl.value = '';
+    } else {
+      input.classList.add('inputError');
+      setTimeout(() => {
+        input.classList.remove('inputError');
+      }, 1000);
+    }
+  });
+};
```
Comment (gemmen29) [inline src/modules/addTask.js]: - [x] Kindly store the task `index` when you add a new task and set the index value with the new array length to achieve this [requirement](https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md#:~:text=By%20default%20new%20tasks%20should%20have%20the%20property%20completed%20set%20to%20false%20and%20the%20property%20index%20set%20to%20the%20value%20of%20the%20new%20array%20length%20(i.e.%20if%20you%27re%20adding%20a%205th%20task%20to%20the%20list%2C%20the%20index%20of%20that%20task%20should%20equal%20to%205).).

File: src/modules/renderTasks.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  const removeTask = () => {
+    const removeBtnsEl = [...document.getElementsByClassName('dltTaskBtn')];
+    removeBtnsEl.forEach((item) => {
+      item.addEventListener('click', (e) => {
+        itemsArray.splice(e.target.id, 1);
+        localStorage.setItem('itemsArray', JSON.stringify(itemsArray));
+        renderToDoItems();
+      });
+    });
+  };
```
Comment (gemmen29) [inline src/modules/renderTasks.js]: - [ ] Please make the delete task update all remaining items' indexes to achieve this [requirement](https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md#:~:text=Deleting%20a%20task%20should%20update%20all%20remaining%20items%27%20indexes%2C%20so%20they%20represent%20the%20current%20list%20order%20and%20are%20unique(i.e.%20if%20you%27re%20deleting%20the%20first%20task%20index%201%20from%20the%20list%2C%20the%20index%20of%20the%20next%20task(2)%20should%20set%20to%201)..).

**Review by gemmen29 — CHANGES_REQUESTED** [review]
Hi @GraceAbike, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 

## Required Changes ♻️
_Check the comments under the review._



## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/addTask.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+const addTask = () => {
+  addElBtn.addEventListener('click', () => {
+    const newTask = todoinputEl.value;
+    if (newTask) {
+      const newTodo = {
+        description: newTask,
+        completed: false,
+      };
+      itemsArray.push(newTodo);
```
Comment (TedLivist) [inline src/modules/addTask.js:16]: - [ ] You are yet to implement the change requested by the previous reviewer. From the requirements, when a task is added it should have these properties: `description`, `index`, and `completed`, but from the image below you only have `description` and `completed`. Kindly include the missing `index` property.

  Also note that when a task is added, its `index` should be set to the length of the array. So for example, if I add the first task, its `index` should be `1`, and when I add the second task, its `index` should be `2`, and so on.

  jff

File: src/modules/renderTasks.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  const removeTask = () => {
+    const removeBtnsEl = [...document.getElementsByClassName('dltTaskBtn')];
+    removeBtnsEl.forEach((item) => {
+      item.addEventListener('click', (e) => {
+        itemsArray.splice(e.target.id, 1);
+        localStorage.setItem('itemsArray', JSON.stringify(itemsArray));
+        renderToDoItems();
+      });
+    });
```
Comment (TedLivist) [inline src/modules/renderTasks.js:36]: - [ ] As requested previously remember to update the indexes of the task when a task is deleted. So for example, if I have **`three`** tasks with the `index` properties: `1`, `2`, `3`, If I delete the second task (with `index: 2`), the remaining tasks should be updated to have index `1` and `2`, not `1` and `3`

  ```
  [{description: 'first task', completed: false, index: 1},
    {description: 'second task', completed: false, index: 2},
    {description: 'third task', completed: false, index: 3}]
  ```
  When I remove second task the new array should be:
  ```
  [{description: 'first task', completed: false, index: 1},
    {description: 'third task', completed: false, index: 2}]
  ```

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @GraceAbike,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by TedLivist — COMMENTED** [review]
## STATUS: INVALID CODE REVIEW REQUEST 🛑 🛑

Hi @GraceAbike,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

### You have not implemented (or pushed) any of the changes from the previous reviews

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: src/modules/addTask.js (hunk 1 above)
Comment (mckent05) [reply src/modules/addTask.js]: - Just like the previous reviewers have suggested, and according to the project requirements, you are required to have an  `index` key when a task is added to your to-do list. I noticed this hasn't been implemented yet. Please its very important you fix this 🙏

File: src/modules/renderTasks.js (hunk 2 above)
Comment (mckent05) [reply src/modules/renderTasks.js]: - Also, like the reviewers have suggested and according to the project requirements,  you are required to update the indexes of each task when a task is deleted. Please ensure you fix this 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @GraceAbike , 

# STATUS : CHANGES REQUIRED ♻️ ♻️ 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your project is looking great 👍 👍 


 ## Required Changes ♻️
- Your project actually looks great and almost all functionalities are working properly, but it is very important you implement changes required by the code reviewers as your project will not be approved otherwise. 🙏 🙏 
_Check the comments under the review._

## Optional suggestions

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your code.


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/renderTasks.js (hunk 2 above)
Comment (GraceAbike) [reply src/modules/renderTasks.js]: @mckent05 thanks for your comments, I don't understand what you mean please 
https://www.webmobilefirst.com/en/screencasts/85OS_II7fL/

image

image

**Review by sja-thedude — CHANGES_REQUESTED** [review]
## Required Changes ♻️

Hi @GraceAbike, This is @sja-thedude your code reviewer.

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

200w

 ## Highlights
✔️ Good job so far

 ## Required Changes ♻️
- [ ] Kindly make sure to implement the changes asked as per the previous reviewer.
1. [here](https://github.com/GraceAbike/To-Do-List/pull/2#discussion_r988598580)
2. [here](https://github.com/GraceAbike/To-Do-List/pull/2#discussion_r988600188)

By default, new tasks should have the property completed set to false and the property index set to the value of the new array length (i.e. if a 5th task is added to the list, the index of that task should equal to 5).
Deleting a task should update all remaining items' indexes, so they represent the current list order and are unique.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [@sja-thedude] in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Qoosim — APPROVED** [review]
#   *Status: Approved* :tada: :heavy_check_mark:

Hi @GraceAbike, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

-   No linter errors. :heavy_check_mark: 
-   Correct GitHub Flow. :heavy_check_mark: 
-   App meet project requirements. :heavy_check_mark:



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**